### ISSUE TITLE
Crash during XLA compilation (and GPU runtime crash)

### ISSUE THREAD

**Issue by lgarrison** [flatironinstitute/jax-finufft#37]
@Matematija has found a bug (or bugs) triggered by the following code:

```python
from functools import partial
import numpy as np
import jax
from jax import numpy as jnp
from jax_finufft import nufft1, nufft2

@jax.jit
@partial(jax.vmap, in_axes=(0, 0, None))
def cconv_test(f, xs, kernel):

    # f.shape = (n_grid, in_features)
    # x.shape = (n_grid, ndim)
    # kernel.shape = (*k_grid, in_features, out_features)

    ndim = xs.shape[-1]
    k_grid_shape = kernel.shape[:-2]

    f_ = f.transpose().astype(complex)
    coords = [xs[...,i] for i in range(ndim)]

    f_hat = nufft1(k_grid_shape, f_, *coords, iflag=-1)
    c_hat = jnp.einsum('a...,...ab->b...', f_hat, kernel)
    return nufft2(c_hat, *coords, iflag=1)

kernel = jnp.array(np.random.randn(32, 32, 32, 16, 16))
f = jnp.array(np.random.randn(8, 100, 16))
x = jnp.array(np.random.uniform(low=-np.pi, high=np.pi, size=(8, 100, 3)))

fconv = cconv_test(f, x, kernel) # crash!
```

It crashes both the CPU and GPU versions, but in very different ways, so I'll focus on CPU for now. The error is:

```console
$ JAX_PLATFORMS=cpu python scripts/jittest.py
[external/com_google_absl/absl/status/statusor.cc : 83] RAW: Attempting to fetch value instead of handling error INTERNAL: Could not convert JSON string to proto: Encountered non UTF-8 code points.
�7�5����
Aborted (core dumped)
```

The error is pretty sensitive to having `nufft2` using the output of `nufft1`, and using `jnp.einsum`, and `vmap`. There are lots of ways to tweak this function that make the crash go away, but it's not clear if doing so is just hiding an underlying problem.

The GDB traceback for this error suggests that the crash happens inside the XLA compilation, before execution even begins. Here's a partial traceback:
```
#0  0x00007ffff66baacf in raise () from /lib64/libc.so.6
#1  0x00007ffff668dea5 in abort () from /lib64/libc.so.6
#2  0x00007fffe2ed3d1c in absl::lts_20230802::raw_log_internal::(anonymous namespace)::RawLogVA(absl::lts_20230802::LogSeverity, char const*, int, char const*, __va_list_tag*) ()
   from /mnt/home/lgarrison/jax-finufft/venv-matija/lib/python3.10/site-packages/jaxlib/xla_extension.so
#3  0x00007fffe2ed3dae in absl::lts_20230802::raw_log_internal::RawLog(absl::lts_20230802::LogSeverity, char const*, int, char const*, ...) () from /mnt/home/lgarrison/jax-finufft/venv-matija/lib/python3.10/site-packages/jaxlib/xla_extension.so
#4  0x00007fffe2e95ffc in absl::lts_20230802::internal_statusor::Helper::Crash(absl::lts_20230802::Status const&) ()
   from /mnt/home/lgarrison/jax-finufft/venv-matija/lib/python3.10/site-packages/jaxlib/xla_extension.so
#5  0x00007fffdb210c7c in xla::cpu::EmitDotOperation(xla::HloInstruction const&, xla::llvm_ir::IrArray const&, xla::llvm_ir::IrArray const&, xla::llvm_ir::IrArray const&, xla::llvm_ir::IrArray const*, llvm::Value*, llvm::IRBuilder<llvm::ConstantFolder, llvm::IRBuilderDefaultInserter>*, mlir::MLIRContext*, xla::HloModuleConfig const&, xla::cpu::TargetMachineFeatures const&) ()
   from /mnt/home/lgarrison/jax-finufft/venv-matija/lib/python3.10/site-packages/jaxlib/xla_extension.so
#6  0x00007fffdb1ed68b in xla::cpu::IrEmitter::HandleDot(xla::HloInstruction*) ()
   from /mnt/home/lgarrison/jax-finufft/venv-matija/lib/python3.10/site-packages/jaxlib/xla_extension.so
#7  0x00007fffe2b98600 in absl::lts_20230802::Status xla::HloInstruction::Visit<xla::HloInstruction*>(xla::DfsHloVisitorBase<xla::HloInstruction*>*) () from /mnt/home/lgarrison/jax-finufft/venv-matija/lib/python3.10/site-packages/jaxlib/xla_extension.so
#8  0x00007fffdacaf158 in absl::lts_20230802::Status xla::HloComputation::AcceptOrdered<xla::HloInstruction*>(xla::DfsHloVisitorBase<xla::HloInstruction*>*, absl::lts_20230802::Span<xla::HloInstruction* const>) const ()
   from /mnt/home/lgarrison/jax-finufft/venv-matija/lib/python3.10/site-packages/jaxlib/xla_extension.so
#9  0x00007fffdb1f439c in xla::cpu::IrEmitter::EmitComputation(xla::HloComputation*, std::basic_string_view<char, std::char_traits<char> >, bool, absl::lts_20230802::Span<xla::HloInstruction* const>, bool) ()
   from /mnt/home/lgarrison/jax-finufft/venv-matija/lib/python3.10/site-packages/jaxlib/xla_extension.so
#10 0x00007fffdb1051ad in xla::cpu::CpuCompiler::CompileLegacyCpuExecutable(std::unique_ptr<xla::HloModule, std::default_delete<xla::HloModule> >) () from /mnt/home/lgarrison/jax-finufft/venv-matija/lib/python3.10/site-packages/jaxlib/xla_extension.so
```

Interestingly, error bisects to ec080e48b69558c7cbb5e798f0bc83f37d1c3b0a, where the XLA translation rules were replaced with MLIR lowering rules. So I think it's possible that a bug was introduced there.

But a counterpoint to that reasoning is that this script also crashes with the CUDA backend, and it still crashes even before that commit! However, the crash is very different than the CPU crash:
```console
$ JAX_PLATFORMS=cuda python scripts/jittest.py
2023-11-17 15:59:54.546455: W external/xla/xla/service/gpu/runtime/support.cc:58] Intercepted XLA runtime error:
UNKNOWN: nufft3d2f XLA extension have thrown an exception: exclusive_scan failed to synchronize: cudaErrorIllegalAddress: an illegal memory access was encountered
2023-11-17 15:59:54.546537: E external/xla/xla/pjrt/pjrt_stream_executor_client.cc:2716] Execution of replica 0 failed: INTERNAL: Failed to execute XLA Runtime executable: run time error: custom call 'xla.gpu.custom_call' failed: nufft3d2f XLA extension have thrown an exception: exclusive_scan failed to synchronize: cudaErrorIllegalAddress: an illegal memory access was encountered; current tracing scope: custom-call.6; current profiling annotation: XlaModule:#hlo_module=jit_cconv_test,program_id=5#.
jax.errors.SimplifiedTraceback: For simplicity, JAX has removed its internal frames from the traceback of the following exception. Set JAX_TRACEBACK_FILTERING=off to include these.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/mnt/home/lgarrison/jax-finufft/scripts/jittest.py", line 51, in <module>
    fconv = cconv_test(f, x, kernel) # crash!
jaxlib.xla_extension.XlaRuntimeError: INTERNAL: Failed to execute XLA Runtime executable: run time error: custom call 'xla.gpu.custom_call' failed: nufft3d2f XLA extension have thrown an exception: exclusive_scan failed to synchronize: cudaErrorIllegalAddress: an illegal memory access was encountered; current tracing scope: custom-call.6; current profiling annotation: XlaModule:#hlo_module=jit_cconv_test,program_id=5#.
```

This appears to be a runtime, GPU-side crash, as opposed to the CPU XLA compilation crash. I can catch the error in `cuda-gdb`, so I'll continue to debug on that front.

So to summarize:
1) Maybe ec080e48b69558c7cbb5e798f0bc83f37d1c3b0a introduced a lowering rules bug that's triggering an XLA compilation crash, and there's also different GPU runtime bug.
2) Maybe the bugs have a common source, and ec080e48b69558c7cbb5e798f0bc83f37d1c3b0a simply moved things around enough to start triggering a CPU bug

In terms of a "common source", I think it would have to be something in the reproducer script (like bad shapes that aren't caught properly) or a JAX bug (e.g. with `einsum`). Because the XLA crash happens so early, I think we could reproduce this outside of jax-finufft, with no compiled library code. I'll have to test this to be sure, though.

@dfm Do you know if a bug in the lowering rules could cause an XLA compilation crash like this?

**Comment by Matematija** [comment]
Oh wow, this is much deeper/weirder than I thought. Thank you for putting so much effort into this.

**Comment by lgarrison** [comment]
In the end, it did turn out that there were two independent bugs here (three if you count the CUDA registers-per-SM problem). That's an impressively destructive reproducer script, @Matematija! Joking aside, it was incredibly helpful to have this minimal reproducer, so thank you for your hard work making it and reporting the issues.
